### PR TITLE
Add message handlers for downstream api auth

### PR DESCRIPTION
**Pull request by qetza**
Authentication message handlers to inject authorization header using token acquisition.

### CODE REVIEW

**Comment by qetza** [comment]
This is for the feature request #1131

File: src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationMessageHandlerOptions.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// Licensed under the MIT License.
+
+using System;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Options passed-in to Microsoft Identity message handlers.
+    /// </summary>
+    public class MicrosoftIdentityAuthenticationMessageHandlerOptions : ICloneable
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationMessageHandlerOptions.cs:11]: @qetza why not use the TokenAcquisitionOptions directly (maybe inheriting from them) ?

File: src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityUserAuthenticationMessageHandler.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            var authResult = await TokenAcquisition.GetAuthenticationResultForUserAsync(
+                options.GetScopes(),
+                authenticationScheme: options.AuthenticationScheme,
+                tenantId: options.Tenant,
+                userFlow: userflow,
+                tokenAcquisitionOptions: options.TokenAcquisitionOptions)
+                .ConfigureAwait(false);
+
+            // add authorization header
+            request.Headers.Add(
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityUserAuthenticationMessageHandler.cs:64]: Are we sure the request won't already have an Authorization header? 
(would we want to use `[ ]` ?

**Review by jmprieur — COMMENTED** [review]
Thanks @qetza 
This looks good to me.

How would you use the message handler?
Would you have a sample app? unit tests? 

Thanks!

**Comment by qetza** [comment]
Hi @jmprieur,
I will add some unit tests and copy an existing downstream webapi sample and rewrite it with using the message handler instead of the DownstreamWebApi

File: src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationMessageHandlerOptions.cs (hunk 1 above)
Comment (qetza) [reply src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationMessageHandlerOptions.cs:11]: @jmprieur, i think you mean reusing `DownstreamWebApiOptions`. This class effectively is a subset of `DownstreamWebApiOptions`, i could make `DownstreamWebApiOptions` inherit from this one and add its specific properties.

File: src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityUserAuthenticationMessageHandler.cs (hunk 2 above)
Comment (qetza) [reply src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityUserAuthenticationMessageHandler.cs:64]: Indexer property `[ ]` is not supported on `HttpHeaders` but i will add a test to remove an existing headers if already present.

**Comment by qetza** [comment]
Hi @jmprieur,
I've refactor the code to have base options class between downstream and delegating handlers, fix the issue with existing header and added some unit tests.
For the sample part as they are in another repo i'll do a PR there later.

**Comment by jmprieur** [comment]
Thanks @qetza

**Comment by qetza** [comment]
I've rebased the code on the lastest version of `master` to resolve merge conflict.

**Comment by jmprieur** [comment]
Sorry @qetza : I missed that you had added tests.
We'll review. thanks for the head-up.

File: src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAppAuthenticationMessageHandler.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// A DelegatingHandler implementation that add an authorization header with a token for the application.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAppAuthenticationMessageHandler.cs:14]: >add [](http://example.com/codeflow?start=48&length=3)

nit: adds

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Comment by jennyf19** [comment]
thanks @qetza i, too, missed the updates here. Could you resolve the merge conflict and then we will merge it? thanks so much.

**Comment by qetza** [comment]
@jennyf19, i've rebased the branch and fixed the merge conflict (on `Microsoft.Identity.Web.xml` as usual)

**Comment by NWessel** [comment]
@qetza Could you provide a link to a repo where you use it?
I really want to try it out right now :)

**Comment by qetza** [comment]
Hi @NWessel,
I've submitted a pull request on the sample repo with an example of using the `MicrosoftIdentityUserAuthenticationMessageHandler` class.

You can see the code here:
https://github.com/qetza/active-directory-aspnetcore-webapp-openidconnect-v2/blob/142a315a36293e25ba5e4fd42640a1a624519d13/3-WebApp-multi-APIs/Startup.cs#L49

**Comment by NWessel** [comment]
Just checked it out, looks like what i would expect it to be ^^
Probably could make the scope optional as well, and take the base url of the request and add /.default

So baseUrl + "/.default"
If no scope is provided.

What is the service name for ? I'm not sure I quite understand the usecase :)

**Comment by qetza** [comment]
You can always create your own extension method without a scope and do the magic in there :smile:

The service name is used internally to link and retrieve the options if you have multiple clients which needs different options.

**Pull request merged**